### PR TITLE
Fix application of optimized iterator for group by queries with a single key

### LINKED ISSUE #15133: Performance regression for group by queries
### CrateDB version

5.5.0

### CrateDB setup information

docker run -e CRATE_HEAP_SIZE=8g --publish 4200:4200 --publish 5432:5432 crate:5.5.0 -Cdiscovery.type=single-node

### Observed behavior

While exploring potential approaches to implement https://github.com/crate/crate/issues/13818, I noticed that since version 5.5.0, there is no significant difference in performance between the following two queries:

```
## Running Query:
   Statement:
     select count(distinct "cCode") from uservisits
   Concurrency: 15
   Iterations: 1000
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1000/1000 [04:20<00:00,  3.84 requests/s]
Runtime (in ms):
    mean:    3873.895 ± 25.701
    min/max: 1175.962 → 5471.169
Percentile:
    50:   3858.094 ± 414.655 (stdev)
    95:   4550.623
    99.9: 5451.031

## Running Query:
   Statement:
     select count("cCode") from (select "cCode" from uservisits group by "cCode") temp
   Concurrency: 15
   Iterations: 1000
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1000/1000 [04:09<00:00,  4.01 requests/s]
Runtime (in ms):
    mean:    3711.499 ± 33.076
    min/max: 610.013 → 6218.592
Percentile:
    50:   3673.261 ± 533.643 (stdev)
    95:   4679.652
    99.9: 6154.636
```

In version 5.4.5, the second query was significantly faster:
```
## Running Query:
   Statement:
     select count(distinct "cCode") from uservisits
   Concurrency: 15
   Iterations: 1000
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1000/1000 [04:22<00:00,  3.81 requests/s]
Runtime (in ms):
    mean:    3893.742 ± 22.845
    min/max: 966.737 → 5314.928
Percentile:
    50:   3894.481 ± 368.590 (stdev)
    95:   4456.188
    99.9: 5096.855

## Running Query:
   Statement:
     select count("cCode") from (select "cCode" from uservisits group by "cCode") temp
   Concurrency: 15
   Iterations: 1000
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1000/1000 [00:16<00:00, 60.13 requests/s]
Runtime (in ms):
    mean:    241.084 ± 3.909
    min/max: 108.970 → 780.635
Percentile:
    50:   239.353 ± 63.063 (stdev)
    95:   316.764
    99.9: 725.190
```

The flamegraphs captured during the execution of the above benchmarks for the query `select count("cCode") from (select "cCode" from uservisits group by "cCode") temp` suggest that in version 5.4.5, the GroupByOptimizedIterator was utilized.

5.4.5:
image

5.5.0:
image

After a brief investigation, I think the issue lies in these lines: https://github.com/crate/crate/blob/a2bfe918d8bcb44f0312b639f649ce10afbe0367/server/src/main/java/io/crate/execution/engine/collect/GroupByOptimizedIterator.java#L135-L138 Currently, for the mentioned query, `keyFieldType` is always null. Consequently, the `GroupByOptimizedIterator` is never applied. This is because `keyRef.column().fqn()` evaluates to `cCode`, and the underlying map storing field types looks like this:
image
And used to look like this:
image

### Steps to Reproduce

Run this benchmark: https://github.com/piotrrzysko/crate-benchmarks/blob/eliminate_distinct/specs/distinct_eliminate.toml.

### PR DESCRIPTION
**Pull request by piotrrzysko**
## Summary of the changes / Why this improves CrateDB

Fixes #15133

## Checklist

 - [x] Added an entry in the latest `docs/appendices/release-notes/<x.y.0>.rst` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in the latest `docs/appendices/release-notes/<x.y.0>.rst`
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: docs/appendices/release-notes/5.5.2.rst [hunk 1]
```diff
@@ -49,4 +49,5 @@ See the :ref:`version_5.5.0` release notes for a full list of changes in the
 Fixes
 =====
 
-None
+- Fixed a performance regression introduced in 5.5.0, ensuring that the
+optimization of group by queries with a single column is applied.
```
Comment (BaurzhanSakhariev) [inline docs/appendices/release-notes/5.5.2.rst:53]:
```suggestion
- Fixed a performance regression introduced in 5.5.0 for queries with
   ``GROUP BY`` on a single column.
```

**Comment by BaurzhanSakhariev** [comment]
ok to test

**Comment by piotrrzysko** [comment]
Here are the benchmark results after running it against this PR:

```
## Running Query:
   Statement:
     select count(distinct "cCode") from uservisits
   Concurrency: 15
   Iterations: 1000
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1000/1000 [04:27<00:00,  3.74 requests/s]
Runtime (in ms):
    mean:    3977.048 ± 27.663
    min/max: 699.374 → 5674.092
Percentile:
    50:   4007.067 ± 446.320 (stdev)
    95:   4613.447
    99.9: 5599.754

## Running Query:
   Statement:
     select count("cCode") from (select "cCode" from uservisits group by "cCode") temp
   Concurrency: 15
   Iterations: 1000
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1000/1000 [00:14<00:00, 71.03 requests/s]
Runtime (in ms):
    mean:    200.548 ± 2.726
    min/max: 34.564 → 307.450
Percentile:
    50:   206.024 ± 43.988 (stdev)
    95:   269.634
    99.9: 307.364
```

File: server/src/testFixtures/java/io/crate/operation/aggregation/AggregationTestCase.java [hunk 2]
```diff
@@ -212,17 +207,14 @@ public Object executeAggregation(Signature maybeUnboundSignature,
             }
         }
         List<Reference> targetColumns = toReference(actualArgumentTypes);
-        var shard = newStartedPrimaryShard(buildMapping(targetColumns));
```
[the remark this answers is not in the document]
Comment (BaurzhanSakhariev) [reply server/src/testFixtures/java/io/crate/operation/aggregation/AggregationTestCase.java:215]: >Both DocValuesGroupByOptimizedIteratorTest and GroupByOptimizedIteratorTest don't extend AggregationTestCase

sorry, I missed that, then change is totally fin.

>My reasoning was that if one test fails before closing a resource, we don't need to care about the fact that the resource might affect other tests, as we need to make the test pass first. 

True, if some test fails, resource closing is the least problem for that particular test. I'm thinking about parallel running multiple tests, some of them failing -> leaking resources --> "good tests" not being able to get a resource --> good tests showing false positive failures.

Maybe it's overthinking but I would vote to align with the only existing usage (wrap to finally) to be on the safe side.

**Review by BaurzhanSakhariev — APPROVED** [review]
Thanks!

**Comment by BaurzhanSakhariev** [comment]
unrelated flaky test

retest this please

**Pull request merged**